packages/vite/src/node/server/index.ts

```diff
@@ -336,6 +336,10 @@ export async function _createServer(
   const config = await resolveConfig(inlineConfig, 'serve')
 
   const { root, server: serverConfig } = config
+  if (!serverConfig?.port) {
```

Comment:
The issue now is that if the user defined the port to something like `3000`, then any restart will do `3001` -> `3000`

Reply:
Emm, the scenario you are referring to is that the user sets the port to 3000, 3000 is already occupied, and the automatic increment becomes 3001. And after modifying the configuration file and restarting, is the 3000 port used again?

Reply:
No in case the first one on 3000 was stop. In that case I think we should preserve `3001`. My previous comment was more about if the user change the port and it trigger a restart, it should be respected

Reply:
Sorry, I'm not sure if there's something wrong with my understanding. Based on your description, it seems that I mean the same thing as the original [comment](https://github.com/vitejs/vite/pull/14418#issue-1904210471). 🤔

Reply:
I agree with your original comment (which I will call case 1)

I will add case 2:
- Start a project with config.server.port set to 3000, but taken so it starts on 3001. 
- Server on port 3000 closes
- Update '.env' file -> server restart on 3001

And case 3:
- Start a project with config.server.port set to 3000, but taken so it starts on 3001. 
- Update config to have server.port set to 4000 -> server restart on 4000
(probably there is an existing test for that one?)

By reading the code, my understanding is that:
Main: Case 1 ❌ Case 2  ❌  Case 3 ✅
First commit: Case 1  ✅ Case 2  ✅  Case 3 ❌
This commit: Case 1  ✅ Case 2  ❌  Case 3  ✅

Reply:
Thanks a lot for the detailed description of the situation, everything should be fine now.

Reply:
Thanks for the update. At first look I though this was too complex but taking time to read it it seems we can't avoid maintaining the two states.

Let's add a comment like this to help people get the why of this code:

```js
// When using non strict port for the dev server, the running port can be different from the config one.
// When restarting, the original port may be available but to avoid a switch of URL for the running 
// browser tabs, we enforce the previously used port, expect if the config port changed.
```